filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+}
+
+func getMtime(path string) (time.Time, error) {
+	stat, err := os.Stat(path)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return stat.ModTime(), nil
+}
+
+func (r *Result[T]) watcherLoop(
+	watcher *fsnotify.Watcher,
+	path string,
+	parser Parser[T],
+	softLimit, hardLimit int64,
+	pollingInterval time.Duration,
+	fsEventsDelay time.Duration,
+) {
+	var lock sync.Mutex
+	forceReload := func() {
+		// make sure we don't run forceReload concurrently
+		lock.Lock()
+		defer lock.Unlock()
+
+		d, mtime, files, err := openAndParse(path, parser, softLimit, hardLimit)
+		if err != nil {
+			slog.ErrorContext(r.ctx, "filewatcher: openAndParse returned error", "err", err)
+		} else {
+			r.data.Store(&atomicData[T]{
+				data:  d,
+				mtime: mtime,
+			})
+			// remove all previously watched files
+			for _, path := range watcher.WatchList() {
+				watcher.Remove(path)
+			}
+			// then read all new files to watch
+			for _, path := range files {
+				if err := watcher.Add(path); err != nil {
+					slog.ErrorContext(r.ctx, "filewatcher: failed to watch file", "err", err, "path", path)
+				}
+			}
+		}
+	}
+
+	reload := func() {
+		mtime, err := getMtime(path)
+		if err != nil {
+			slog.ErrorContext(r.ctx, "filewatcher: failed to get mtime for file", "err", err, "path", path)
+			return
+		}
+		if r.data.Load().(*atomicData[T]).mtime.Before(mtime) {
+			forceReload()
+		}
+	}
+
+	var tickerChan <-chan time.Time
+	if pollingInterval > 0 {
+		ticker := time.NewTicker(pollingInterval)
+		defer ticker.Stop()
+		tickerChan = ticker.C
+	}
+	var timer *time.Timer
+	for {
+		select {
+		case <-r.ctx.Done():
+			watcher.Close()
+			return
+
+		case err := <-watcher.Errors:
+			slog.ErrorContext(r.ctx, "filewatcher: watcher error", "err", err)
+
+		case ev := <-watcher.Events:
+			// When both r.ctx.Done() and watcher.Events are unblocked, there's no
+			// guarantee which case would be picked, so do an additional ctx check
+			// here to make sure we don't spam the log with i/o errors (which mainly
+			// happen in tests)
+			if r.ctx.Err() != nil {
+				continue
+			}
+
+			switch ev.Op {
+			default:
+				// Ignore uninterested events.
+			case fsnotify.Create, fsnotify.Write, fsnotify.Rename, fsnotify.Remove:
+				// Use fsEventDelay to avoid calling forceReload repetively when a burst
+				// of fs events happens (for example, when multiple files within the
+				// directory changed).
+				if timer == nil {
+					timer = time.AfterFunc(fsEventsDelay, forceReload)
+				} else {
+					timer.Reset(fsEventsDelay)
+				}
+			}
+
+		case <-tickerChan:
+			// When both r.ctx.Done() and tickerChan are unblocked, there's no
+			// guarantee which case would be picked, so do an additional ctx check
+			// here to make sure we don't spam the log with i/o errors (which mainly
+			// happen in tests)
+			if r.ctx.Err() != nil {
+				continue
+			}
+
+			reload()
+		}
+	}
+}
+
+// The actual data held in Result.data.
+type atomicData[T any] struct {
```

Comment:
the name `atomicData` to me implies a type that handles the atomicity itself, rather than the data stored inside an atomic.Value.

I would probably call it `dataSnapshot` or something to imply that it's the pairing of a data and its mtime

Reply:
ok `dataAt` it is then :)